File: public/sdk/node/expense-tracker.js
/**
 * FounderHQ Expense Tracker - Node.js SDK
 * Track expenses from custom integrations and AI APIs
 */

class ExpenseTracker {
  constructor(apiKey, baseUrl = "https://api.founderhq.com") {
    this.apiKey = apiKey
    this.baseUrl = baseUrl
  }

  async track(options) {
    const { service, amount, category = "Third-Party Tools", pricingModel = "usage", meta = {} } = options

    const payload = {
      service,
      amountUsd: amount,
      category,
      pricingModel,
      meta,
      timestamp: Date.now(),
    }

    const response = await fetch(`${this.baseUrl}/v1/expenses`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify(payload),
    })

    return response.json()
  }
}

/**
 * Wrap OpenAI client to track token usage
 */
export function trackOpenAI(openaiClient, options) {
  const { apiKey, serviceName = "OpenAI GPT-4", costPer1kTokens = 0.03 } = options
  const tracker = new ExpenseTracker(apiKey)

  return new Proxy(openaiClient, {
    get(target, prop) {
      const original = target[prop]

      if (prop === "chat") {
        return new Proxy(original, {
          get(chatTarget, chatProp) {
            const chatOriginal = chatTarget[chatProp]

            if (chatProp === "completions") {
              return new Proxy(chatOriginal, {
                get(completionsTarget, completionsProp) {
                  const completionsOriginal = completionsTarget[completionsProp]

                  if (completionsProp === "create") {
                    return async (...args) => {
                      const result = await completionsOriginal.apply(completionsTarget, args)

                      // Track token usage
                      if (result.usage) {
                        const totalTokens = result.usage.total_tokens
                        const cost = (totalTokens / 1000) * costPer1kTokens

                        await tracker.track({
                          service: serviceName,
                          amount: cost,
                          category: "AI Tokens / APIs",
                          pricingModel: "usage",
                          meta: { tokens: totalTokens },
                        })
                      }

                      return result
                    }
                  }

                  return completionsOriginal
                },
              })
            }

            return chatOriginal
          },
        })
      }

      return original
    },
  })
}

/**
 * Create an expense tracker for custom usage
 */
export function trackExpense(options) {
  const { apiKey, serviceName, category = "Third-Party Tools", pricingModel = "usage" } = options

  const tracker = new ExpenseTracker(apiKey)

  return {
    async track(expenseData) {
      const { cost, meta = {} } = expenseData

      return tracker.track({
        service: serviceName,
        amount: cost,
        category,
        pricingModel,
        meta,
      })
    },
  }
}

// Example usage
async function example() {
  // Example 1: Track OpenAI usage
  const OpenAI = (await import("openai")).default
  const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY })

  const trackedOpenAI = trackOpenAI(openai, {
    apiKey: "your-founderhq-api-key",
    serviceName: "OpenAI GPT-4",
  })

  const response = await trackedOpenAI.chat.completions.create({
    model: "gpt-4",
    messages: [{ role: "user", content: "Hello!" }],
  })

  // Example 2: Track custom expense
  const customTracker = trackExpense({
    apiKey: "your-founderhq-api-key",
    serviceName: "Custom AI API",
    category: "AI Tokens / APIs",
  })

  await customTracker.track({
    cost: 0.05,
    meta: { requests: 1, tokens: 1500 },
  })
}

export { ExpenseTracker }

File: public/sdk/python/expense_tracker.py
"""
FounderHQ Expense Tracker - Python SDK
Track expenses from custom integrations and AI APIs
"""

import functools
import time
from typing import Any, Callable, Dict, Optional
import requests


class ExpenseTracker:
    """Main expense tracker client"""
    
    def __init__(self, api_key: str, base_url: str = "https://api.founderhq.com"):
        self.api_key = api_key
        self.base_url = base_url
    
    def track(
        self,
        service: str,
        amount: float,
        category: str = "Third-Party Tools",
        pricing_model: str = "usage",
        meta: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Track an expense event"""
        payload = {
            "service": service,
            "amountUsd": amount,
            "category": category,
            "pricingModel": pricing_model,
            "meta": meta or {},
            "timestamp": time.time()
        }
        
        response = requests.post(
            f"{self.base_url}/v1/expenses",
            json=payload,
            headers={"Authorization": f"Bearer {self.api_key}"}
        )
        
        return response.json()


def track_openai_tokens(
    api_key: str,
    service_name: str = "OpenAI GPT-4",
    cost_per_1k_tokens: float = 0.03
) -> Callable:
    """
    Decorator to track OpenAI API token usage
    
    Usage:
        @track_openai_tokens(api_key="your-api-key")
        def generate_text(prompt: str) -> str:
            # Your OpenAI API call
            pass
    """
    tracker = ExpenseTracker(api_key)
    
    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            result = func(*args, **kwargs)
            
            # Extract token usage from OpenAI response
            if hasattr(result, 'usage'):
                total_tokens = result.usage.total_tokens
                cost = (total_tokens / 1000) * cost_per_1k_tokens
                
                tracker.track(
                    service=service_name,
                    amount=cost,
                    category="AI Tokens / APIs",
                    pricing_model="usage",
                    meta={"tokens": total_tokens}
                )
            
            return result
        
        return wrapper
    
    return decorator


def track_anthropic_tokens(
    api_key: str,
    service_name: str = "Anthropic Claude",
    cost_per_1k_tokens: float = 0.025
) -> Callable:
    """
    Decorator to track Anthropic API token usage
    
    Usage:
        @track_anthropic_tokens(api_key="your-api-key")
        def generate_text(prompt: str) -> str:
            # Your Anthropic API call
            pass
    """
    tracker = ExpenseTracker(api_key)
    
    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            result = func(*args, **kwargs)
            
            # Extract token usage from Anthropic response
            if hasattr(result, 'usage'):
                total_tokens = result.usage.input_tokens + result.usage.output_tokens
                cost = (total_tokens / 1000) * cost_per_1k_tokens
                
                tracker.track(
                    service=service_name,
                    amount=cost,
                    category="AI Tokens / APIs",
                    pricing_model="usage",
                    meta={"tokens": total_tokens}
                )
            
            return result
        
        return wrapper
    
    return decorator


def track_expense(
    api_key: str,
    service_name: str,
    category: str = "Third-Party Tools",
    pricing_model: str = "usage"
) -> Callable:
    """
    Generic decorator to track custom expenses
    
    Usage:
        @track_expense(
            api_key="your-api-key",
            service_name="Custom API",
            category="AI Tokens / APIs"
        )
        def call_api(data: dict) -> dict:
            # Your API call
            return {"result": ..., "cost": 0.05, "meta": {"requests": 1}}
    """
    tracker = ExpenseTracker(api_key)
    
    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        def wrapper(*args, **kwargs):
            result = func(*args, **kwargs)
            
            # Extract cost from function return value
            if isinstance(result, dict) and "cost" in result:
                cost = result["cost"]
                meta = result.get("meta", {})
                
                tracker.track(
                    service=service_name,
                    amount=cost,
                    category=category,
                    pricing_model=pricing_model,
                    meta=meta
                )
            
            return result
        
        return wrapper
    
    return decorator


# Example usage
if __name__ == "__main__":
    # Example 1: Track OpenAI usage
    @track_openai_tokens(api_key="your-founderhq-api-key")
    def generate_with_gpt4(prompt: str):
        # Simulated OpenAI response
        class MockResponse:
            class Usage:
                total_tokens = 1500
            usage = Usage()
        
        return MockResponse()
    
    # Example 2: Track custom expense
    @track_expense(
        api_key="your-founderhq-api-key",
        service_name="Custom AI API",
        category="AI Tokens / APIs"
    )
    def call_custom_api(data: dict):
        # Your API logic here
        return {
            "result": {"text": "Generated text"},
            "cost": 0.05,
            "meta": {"requests": 1, "tokens": 1500}
        }
